Matter TemperatureLevel feature from TemperatureControl cluster

lboue opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add support for `TemperatureLevel` feature from Matter [Temperature Control Cluster](https://github.com/project-chip/connectedhomeip/blob/master/src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml)  (0x0056) used in the following device types: 
* [Dishwasher](https://github.com/project-chip/connectedhomeip/blob/79c1cf5cb093448d1eb6ee9b5f31c68b75c241c8/data_model/1.3/device_types/Dishwasher.xml)
* [LaundryDryer](https://github.com/project-chip/connectedhomeip/blob/79c1cf5cb093448d1eb6ee9b5f31c68b75c241c8/data_model/1.3/device_types/LaundryDryer.xml)
* [LaundryWasher](https://github.com/project-chip/connectedhomeip/blob/79c1cf5cb093448d1eb6ee9b5f31c68b75c241c8/data_model/1.3/device_types/LaundryWasher.xml)

You can select the Temperature Level from a select list.

**TemperatureLevel** feature: 
       For  devices that use vendor-specific temperature levels for the temperature setpoint, such as some washers (Dishwasher, LaundryDryer, LaundryWasher)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

![image](https://github.com/user-attachments/assets/67d6dc58-862c-4644-a6ca-3d91425f5bb1)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

lboue committed:
SelectedTemperatureLevel

lboue committed:
fix description

lboue committed:
Fix LaundryWasher fixture

lboue committed:
Tests for TemperatureControl

lboue committed:
Tests part 2

lboue committed:
Tests part 3

lboue commented:
I've split the implementation into 2 entities:

| Device Type                    | Cluster            | Feature           | Map              |
|--------------------------------|--------------------|-------------------|------------------|
| Temperature Controlled Cabinet | TemperatureControl | TemperatureNumber | Platform.CLIMATE |
| Temperature Controlled Cabinet | TemperatureControl |  TemperatureLevel |  Platform.SELECT |

For TemperatureNumber, I managed to make the [Silabs Refrigerator and Temperature Controlled Example](https://github.com/project-chip/connectedhomeip/tree/master/examples/refrigerator-app/silabs) work.

The other PR:
- #128838

lboue committed:
Add test in conftest

lboue committed:
Add test_sensor

marcelveldt commented on homeassistant/components/matter/select.py:
```diff
@@ -111,6 +111,35 @@ def _update_from_device(self) -> None:
             self._attr_name = desc
 
 
+class MatterLevelSelectEntity(MatterSelectEntity):
```
I think we can further generalize this one.
In essence its executing a command with a value. 
So we can have a schema that takes the command

lboue replied:
I'm not sure how to make this more generic, given that there are attribute and command names in this class. Do you have a suggestion?

lboue replied:
I haven't figured out how to make the command argument generic. When the command is called, the variable name is used, not its value.
Do you have a suggestion?

marcelveldt replied:
Refactored the code a bit - now its universal

lboue committed:
secondary_attribute

lboue committed:
More generic

lboue committed:
Make it more generic

lboue committed:
Revert to fix mypy error

Error: homeassistant/components/matter/select.py:130: error: Unexpected keyword argument "targetTemperatureLevel"  [call-arg]
Error: homeassistant/components/matter/select.py:130: error: "None" not callable  [misc]

lboue committed:
ruff-format

lboue committed:
ruff-format

marcelveldt commented on homeassistant/components/matter/models.py:
```diff
@@ -59,6 +59,11 @@ def primary_attribute(self) -> type[ClusterAttributeDescriptor]:
         """Return Primary Attribute belonging to the entity."""
         return self.attributes_to_watch[0]
 
+    @property
+    def secondary_attribute(self) -> type[ClusterAttributeDescriptor]:
+        """Return Secondary Attribute belonging to the entity."""
+        return self.attributes_to_watch[1]
+
```
this is dangerous - lets not do this

marcelveldt commented on homeassistant/components/matter/select.py:
```diff
@@ -254,4 +289,18 @@ def _update_from_device(self) -> None:
+            key="TemperatureControlSelectedTemperatureLevel",
+            translation_key="temperature_level",
+            command=clusters.TemperatureControl.Commands.SetTemperature,
+            # field=clusters.TemperatureControl.Commands.SetTemperature.targetTemperatureLevel,
```
whats this ?

lboue replied:
I'm trying to make this method generic as requested and pass the command and the name of the parameter to be updated as arguments.

marcelveldt replied:
yeah either that or we register a callback for the entire command ?

marcelveldt commented on homeassistant/components/matter/select.py:
```diff
@@ -111,6 +115,37 @@ def _update_from_device(self) -> None:
+        """Change the selected option."""
+        # select the ID from the label string
+        option_id = self._attr_options.index(option)
+        # field = self.entity_description.field
```
?

marcelveldt commented on homeassistant/components/matter/select.py:
```diff
@@ -46,6 +47,9 @@ async def async_setup_entry(
     """Describe Matter select entities."""
 
+    command: Callable[[], Any] | None = None
+    field: str | None = None
```
Is this leftover ?

marcelveldt committed:
revert secondary attribute property

marcelveldt committed:
Refactor select entity classes

Rename MatterAttributeSelectEntity: Make it more clear that MatterAttributeSelectEntity is tied to a single attribute

Add MatterListSelectEntityDescription to define how the MatterListSelectEntity should behave

marcelveldt committed:
Update select.py

marcelveldt committed:
Update select.py

lboue commented:
Now tests are green. It should be OK to validate.

marcelveldt commented on tests/components/matter/test_select.py:
```diff
@@ -103,3 +103,39 @@ async def test_attribute_select_entities(
+
+
+@pytest.mark.parametrize("node_fixture", ["silabs_laundrywasher"])
+async def test_level_select_entities(
```
```suggestion
async def test_list_select_entities(
```

marcelveldt commented on tests/components/matter/test_select.py:
```diff
@@ -103,3 +103,39 @@ async def test_attribute_select_entities(
+    matter_client: MagicMock,
+    matter_node: MatterNode,
+) -> None:
+    """Test select entities are created for the TemperatureControl cluster attributes."""
```
```suggestion
    """Test ListSelect entities are discovered and working, by using a laundrywasher fixture."""
```

marcelveldt committed:
Apply suggestions from code review

marcelveldt reviewed (approved):
Nice, thanks @lboue !

marcelveldt committed:
Merge branch 'dev' into TemperatureLevel

lboue committed:
MatterSelectEntity replaced by MatterAttributeSelectEntity

lboue committed:
Test invalid value

lboue committed:
Merge branch 'dev' into TemperatureLevel

marcelveldt reviewed (approved):
Nice, Thanks @lboue !
